File: src/101_200/0199_binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
	def rightSideView(self, root):
		"""
		:type root: TreeNode
		:rtype: List[int]
		"""
		if(root==None):
			return []
		res = []
		queue = []
		queue.append(root)
		while(queue):
			res.append(queue[-1].val)
			size = len(queue)
			for i in range(size):
				tmp = queue.pop(0)
				if(tmp.left != None):
					queue.append(tmp.left)
				if(tmp.right != None):
					queue.append(tmp.right)
		return res
		
		
	# 另一种解法 BFS方式，遍历节点后将用dict记录最后一个节点和深度
	def rightSideView(self, root):
		if not root:
			return []
		max_depth = -1
		queue = [(root,0)]
		right_view_map = dict()
		while queue:
			node,depth = queue.pop(0)
			max_depth = max(max_depth,depth)
			if node:
				right_view_map[depth] = node.val
				queue.append((node.left,depth+1))
				queue.append((node.right,depth+1))
		ans = []
		for v in right_view_map.values():
			ans.append(v)
		return ans
		
		
		
	# 跟解法二类似，只是改成stack了
	def rightSideView(self, root):		
		if not root:
			return []
		stack = [(root,0)]
		max_depth = -1
		right_view_map = dict()
		while stack:
			node,depth = stack.pop()
			if node:
				max_depth = max(max_depth,depth)
				if depth not in right_view_map:
					right_view_map[depth] = node.val
				stack.append((node.left,depth+1))
				stack.append((node.right,depth+1))
		return [right_view_map[depth] for depth in xrange(max_depth+1)]


		
	# 另一种DFS 实现，很巧妙
	def rightSideView(self, root):		
		ans = []
		def dfs(root,depth):
			if not root:
				return
			if depth==len(ans):
				ans.append(root.val)
			dfs(root.right,depth+1)
			dfs(root.left, depth+1)
		dfs(root,0)
		return ans		
		
		

File: src/101_200/0199_binary-tree-right-side-view/binary-tree-right-side-view.java

//解法一
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        if(root==null) {
            return new ArrayList<Integer>();
        }
        List<Integer> res = new ArrayList<Integer>();
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);
        while(!queue.isEmpty()) {
            int size = queue.size();
            LinkedList<Integer> tmp = new LinkedList<Integer>();
            for(int i=0;i<size;++i) {
                TreeNode node = queue.poll();
                if(node.left!=null) {
                    queue.offer(node.left);
                }
                if(node.right!=null) {
                    queue.offer(node.right);
                }
                tmp.add(node.val);
            }
            res.add(tmp.pollLast());
        }
        return res;
    }
}



//解法二
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        if(root==null) {
            return new ArrayList<Integer>();
        }
		Queue<Pair<Integer,TreeNode>> queue = new LinkedList<Pair<Integer,TreeNode>>();
		Map<Integer,Integer> rightViewMap = new HashMap<Integer,Integer>();
		int maxDepth = Integer.MIN_VALUE;
		queue.offer(new Pair<Integer, TreeNode>(0,root));
		List<Integer> ans = new ArrayList<Integer>();
		while(!queue.isEmpty()) {
			Pair<Integer,TreeNode> pair = queue.poll();
			int depth = pair.key;
			TreeNode node = pair.value;
			if(node!=null) {
				maxDepth = Math.max(maxDepth,depth);
				rightViewMap.put(depth,node.val);
				queue.offer(new Pair<Integer, TreeNode>(depth+1,node.left));
				queue.offer(new Pair<Integer, TreeNode>(depth+1,node.right));
			}
		}
		for(int i=0;i<=maxDepth;++i) {
			ans.add(rightViewMap.get(i));
		}
        return ans;
    }
	
	static class Pair<K,V> {
		final K key;
		final V value;
		Pair(K key,V value) {
			this.key = key;
			this.value = value;
		}
		
		public K getKey() {
			return this.key;
		}
		
		public V getValue() {
			return this.value;
		}
	}
}



